File: enve/Services/Vocabulary/DefinitionLookupService.swift
import Foundation
import Logging

@MainActor
final class DefinitionLookupService {
    static let shared = DefinitionLookupService()

    private var memoryCache: [String: String?] = [:]
    private var inflight: [String: Task<String?, Never>] = [:]

    private let cacheRoot: URL

    private init() {
        let appSupport =
            (try? FileManager.default.url(
                for: .applicationSupportDirectory,
                in: .userDomainMask,
                appropriateFor: nil,
                create: true
            )) ?? URL(fileURLWithPath: NSTemporaryDirectory())
        cacheRoot = appSupport.appendingPathComponent("Enve/DefinitionCache", isDirectory: true)
        try? FileManager.default.createDirectory(at: cacheRoot, withIntermediateDirectories: true)
    }

    func definition(for rawWord: String, language: String? = nil) async -> String? {
        let word = normalize(rawWord)
        guard !word.isEmpty else { return nil }
        let lang = languageTag(from: language)
        let key = "\(lang):\(word)"

        if let cached = memoryCache[key] { return cached }
        if let task = inflight[key] { return await task.value }

        if let starDictHit = formatStarDict(InstalledDictionariesStore.shared.lookup(word)) {
            memoryCache[key] = starDictHit
            writeDisk(lang: lang, word: word, definition: starDictHit)
            return starDictHit
        }

        if let onDisk = readDisk(lang: lang, word: word) {
            memoryCache[key] = onDisk.value
            return onDisk.value
        }

        let task = Task { [lang, word] in
            await fetchOnline(word: word, lang: lang)
        }
        inflight[key] = task
        let result = await task.value
        inflight[key] = nil
        memoryCache[key] = result
        writeDisk(lang: lang, word: word, definition: result)
        return result
    }

    func cached(for rawWord: String, language: String? = nil) -> String? {
        let word = normalize(rawWord)
        let lang = languageTag(from: language)
        if let cached = memoryCache["\(lang):\(word)"] { return cached }
        return readDisk(lang: lang, word: word)?.value
    }

    private func normalize(_ s: String) -> String {
        s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    private func languageTag(from raw: String?) -> String {
        guard let raw, !raw.isEmpty else { return "en" }
        return String(raw.prefix(2)).lowercased()
    }

    private struct DiskResult { let value: String? }

    private func cacheURL(lang: String, word: String) -> URL {
        let dir = cacheRoot.appendingPathComponent(lang, isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let safe = word.replacingOccurrences(of: "/", with: "_")
        return dir.appendingPathComponent("\(safe).txt", isDirectory: false)
    }

    private func readDisk(lang: String, word: String) -> DiskResult? {
        let url = cacheURL(lang: lang, word: word)
        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        let data = (try? Data(contentsOf: url)) ?? Data()
        if data.isEmpty { return DiskResult(value: nil) }
        return DiskResult(value: String(data: data, encoding: .utf8))
    }

    private func writeDisk(lang: String, word: String, definition: String?) {
        let url = cacheURL(lang: lang, word: word)
        let data = (definition ?? "").data(using: .utf8) ?? Data()
        try? data.write(to: url, options: .atomic)
    }

    private struct DictionaryEntry: Decodable {
        let meanings: [Meaning]?
        struct Meaning: Decodable {
            let partOfSpeech: String?
            let definitions: [Definition]?
            struct Definition: Decodable { let definition: String? }
        }
    }

    private func fetchOnline(word: String, lang: String) async -> String? {
        guard let escaped = word.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed),
            let url = URL(string: "https://api.dictionaryapi.dev/api/v2/entries/\(lang)/\(escaped)")
        else { return nil }

        var request = URLRequest(url: url)
        request.timeoutInterval = 8
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                return nil
            }
            let entries = try JSONDecoder().decode([DictionaryEntry].self, from: data)
            return formatDefinition(from: entries)
        } catch {
            AppLogger.network.warning("DefinitionLookupService fetch failed for \(word): \(error.localizedDescription)")
            return nil
        }
    }

    private func formatStarDict(_ hit: StarDictDictionary.Definition?) -> String? {
        guard let hit, !hit.definition.isEmpty else { return nil }

        let raw = hit.definition
        let cleaned: String
        switch hit.type {
        case "h", "g":
            cleaned = stripMarkup(raw)
        default:
            cleaned = raw
        }
        let trimmed = cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private func stripMarkup(_ s: String) -> String {

        let tagless = s.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
        let entitiesDecoded =
            tagless
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&apos;", with: "'")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&nbsp;", with: " ")
        let collapsed = entitiesDecoded.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        return collapsed
    }

    private func formatDefinition(from entries: [DictionaryEntry]) -> String? {
        var lines: [String] = []
        var seen = Set<String>()
        outer: for entry in entries {
            for meaning in entry.meanings ?? [] {
                guard
                    let first = meaning.definitions?.first?.definition?
                        .trimmingCharacters(in: .whitespacesAndNewlines),
                    !first.isEmpty
                else { continue }
                let pos = meaning.partOfSpeech?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                let line = pos.isEmpty ? first : "(\(pos)) \(first)"
                if seen.insert(line).inserted {
                    lines.append(line)
                    if lines.count >= 3 { break outer }
                }
            }
        }
        return lines.isEmpty ? nil : lines.joined(separator: "\n")
    }
}

File: android/engine/src/main/java/com/enve/app/data/vocab/DefinitionLookupService.kt
package com.enve.app.data.vocab

import android.content.Context
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DefinitionLookupService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val installedDictionaries: InstalledDictionariesStore,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val memoryCache = mutableMapOf<String, String?>()
    private val inflight = mutableMapOf<String, Deferred<String?>>()
    private val mutex = Mutex()

    private val client by lazy {
        OkHttpClient.Builder()
            .connectTimeout(8, TimeUnit.SECONDS)
            .readTimeout(8, TimeUnit.SECONDS)
            .build()
    }

    private val json = Json { ignoreUnknownKeys = true }

    private val cacheRoot: File by lazy {
        File(context.filesDir, "DefinitionCache").apply { mkdirs() }
    }

    suspend fun definition(rawWord: String, language: String? = null): String? {
        val word = normalize(rawWord)
        if (word.isEmpty()) return null
        val lang = languageTag(language)
        val key = "$lang:$word"

        mutex.withLock {
            memoryCache[key]?.let { return it }
            inflight[key]?.let { return it.await() }
        }

        formatStarDict(installedDictionaries.lookup(word))?.let { result ->
            mutex.withLock { memoryCache[key] = result }
            writeDisk(lang, word, result)
            return result
        }

        val onDisk = readDisk(lang, word)
        if (onDisk != null) {
            mutex.withLock { memoryCache[key] = onDisk.value }
            return onDisk.value
        }

        val deferred = scope.async {
            fetchOnline(word, lang)
        }
        mutex.withLock { inflight[key] = deferred }
        val result = try { deferred.await() } finally {
            mutex.withLock { inflight.remove(key) }
        }
        mutex.withLock { memoryCache[key] = result }
        writeDisk(lang, word, result)
        return result
    }

    fun cached(rawWord: String, language: String? = null): String? {
        val word = normalize(rawWord)
        val lang = languageTag(language)
        memoryCache["$lang:$word"]?.let { return it }
        return readDisk(lang, word)?.value
    }

    private fun normalize(s: String): String = s.trim().lowercase()

    private fun languageTag(raw: String?): String =
        raw?.take(2)?.lowercase()?.takeIf { it.isNotEmpty() } ?: "en"

    private data class DiskResult(val value: String?)

    private fun cacheFile(lang: String, word: String): File {
        val dir = File(cacheRoot, lang).apply { mkdirs() }
        val safe = word.replace('/', '_')
        return File(dir, "$safe.txt")
    }

    private fun readDisk(lang: String, word: String): DiskResult? {
        val file = cacheFile(lang, word)
        if (!file.exists()) return null
        val bytes = runCatching { file.readBytes() }.getOrDefault(ByteArray(0))
        return if (bytes.isEmpty()) DiskResult(null) else DiskResult(String(bytes))
    }

    private fun writeDisk(lang: String, word: String, definition: String?) {
        runCatching {
            cacheFile(lang, word).writeBytes((definition ?: "").toByteArray())
        }
    }

    @Serializable
    private data class ApiEntry(val meanings: List<ApiMeaning> = emptyList())

    @Serializable
    private data class ApiMeaning(
        val partOfSpeech: String? = null,
        val definitions: List<ApiDefinition> = emptyList(),
    )

    @Serializable
    private data class ApiDefinition(val definition: String? = null)

    private suspend fun fetchOnline(word: String, lang: String): String? = withContext(Dispatchers.IO) {
        val url = "https://api.dictionaryapi.dev/api/v2/entries/$lang/${java.net.URLEncoder.encode(word, "UTF-8")}"
        val request = Request.Builder().url(url).header("Accept", "application/json").build()
        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@use null
                val body = response.body?.string() ?: return@use null
                val entries = json.decodeFromString<List<ApiEntry>>(body)
                format(entries)
            }
        } catch (e: Exception) {
            Log.w(TAG, "DefinitionLookup fetch failed for '$word': ${e.message}")
            null
        }
    }

    private fun formatStarDict(hit: StarDictDictionary.Definition?): String? {
        if (hit == null || hit.definition.isEmpty()) return null
        val cleaned = when (hit.type) {
            'h', 'g' -> stripMarkup(hit.definition)
            else -> hit.definition
        }
        return cleaned.trim().takeIf { it.isNotEmpty() }
    }

    private fun stripMarkup(s: String): String {
        val tagless = s.replace(Regex("<[^>]+>"), "")
        val decoded = tagless
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&apos;", "'")
            .replace("&#39;", "'")
            .replace("&nbsp;", " ")
        return decoded.replace(Regex("\\s+"), " ")
    }

    private fun format(entries: List<ApiEntry>): String? {
        val lines = mutableListOf<String>()
        val seen = mutableSetOf<String>()
        outer@ for (entry in entries) {
            for (meaning in entry.meanings) {
                val first = meaning.definitions.firstOrNull()?.definition?.trim().orEmpty()
                if (first.isEmpty()) continue
                val pos = meaning.partOfSpeech?.trim().orEmpty()
                val line = if (pos.isEmpty()) first else "($pos) $first"
                if (seen.add(line)) {
                    lines.add(line)
                    if (lines.size >= 3) break@outer
                }
            }
        }
        return lines.takeIf { it.isNotEmpty() }?.joinToString("\n")
    }

    companion object { private const val TAG = "DefinitionLookup" }
}
